migrations/multiple_contract_migration.go

```diff
@@ -0,0 +1,230 @@
+		_, err := fmt.Fprintf(&b, "\nerr %d: %s", i, err)
+		if err != nil {
+			panic(err)
+		}
+	}
+	return b.String()
+}
+
+type MultipleContractMigration struct {
+	// the cache keeps a record of already migrated addresses
+	// it is completely ephemeral. If lost, it will only effect performance
+	cache                 map[string]struct{}
+	logger                zerolog.Logger
+	ExceptionalMigrations map[string]func(ledger.Payload, zerolog.Logger) ([]ledger.Payload, error)
+}
+
+func NewMultipleContractMigration(logger zerolog.Logger) MultipleContractMigration {
+	return MultipleContractMigration{
+		cache:                 make(map[string]struct{}, 0),
+		logger:                logger,
+		ExceptionalMigrations: make(map[string]func(ledger.Payload, zerolog.Logger) ([]ledger.Payload, error), 0),
+	}
+}
+
+func (m *MultipleContractMigration) Migrate(payload []ledger.Payload) ([]ledger.Payload, error) {
+	migratedPayloads := make([]ledger.Payload, 0)
+	errors := make([]error, 0)
+
+	for _, p := range payload {
+		results, err := m.migrateRegister(p)
+		// dont fail fast... try to collect errors so multiple errors can be addressed at once
+		if err != nil {
+			errors = append(errors, err)
+			continue
+		}
+		migratedPayloads = append(migratedPayloads, results...)
+	}
+	if len(errors) != 0 {
+		return nil, &MultipleContractMigrationError{
+			Errors: errors,
+		}
+	}
+	return migratedPayloads, nil
+}
+
+func keyToRegisterId(key ledger.Key) (flow.RegisterID, error) {
+	if len(key.KeyParts) != 3 ||
+		key.KeyParts[0].Type != state.KeyPartOwner ||
+		key.KeyParts[1].Type != state.KeyPartController ||
+		key.KeyParts[2].Type != state.KeyPartKey {
+		return flow.RegisterID{}, fmt.Errorf("key not in expected format %s", key.String())
+	}
+
+	return flow.NewRegisterID(string(key.KeyParts[0].Value), string(key.KeyParts[1].Value), string(key.KeyParts[2].Value)), nil
+}
+
+func (m *MultipleContractMigration) migrateRegister(p ledger.Payload) ([]ledger.Payload, error) {
+	registerId, err := keyToRegisterId(p.Key)
+	if err != nil {
+		return nil, err
+	}
+	// have we migrated this?
+	if _, alreadyMigrated := m.cache[registerId.Owner]; alreadyMigrated {
+		// yes, return it as is
+		return []ledger.Payload{p}, nil
+	}
+	if !needsMigration(registerId) {
+		return []ledger.Payload{p}, nil
+	}
+
+	// we are going to migrate this
+	m.cache[registerId.Owner] = struct{}{}
+
+	if em, hasEM := m.ExceptionalMigrations[registerId.Owner]; hasEM {
+		m.logger.Info().
+			Err(err).
+			Str("address", flow.BytesToAddress([]byte(registerId.Owner)).HexWithPrefix()).
+			Msg("Using exceptional migration for address")
+		return em(p, m.logger)
+	}
+
+	return m.migrateValue(p)
+}
+
+func (m *MultipleContractMigration) migrateValue(p ledger.Payload) ([]ledger.Payload, error) {
+
+	// we don't need the the empty code register
+	value := p.Value
+	address := flow.BytesToAddress(p.Key.KeyParts[0].Value)
+	if len(value) == 0 {
+		return []ledger.Payload{}, nil
```

Comment:
This case should never happen but if happens are we sure we want to remove the keys that don't have values?

Reply:
otherwise we should do `[]ledger.Payload{p}`

Reply:
This is actually a very common case. Previously, the account creation code would create an empty code register. This empty code register is not necessary so we just leave it out.